main.go

```diff
@@ -37,58 +26,74 @@ func main() {
 		panic(err)
 	}
 
-	loggerConfig := zap.NewProductionConfig()
-	if err := configureLogLevel(&loggerConfig, cfg.LogLevel); err != nil {
-		panic(fmt.Sprintf("failed to parse log level: %s", err.Error()))
+	log := configureLogger(cfg)
+	m := metrics.NewClient(cfg.Prometheus.Path, cfg.Prometheus.Port, log)
+	go m.RegisterHTTPHandler()
+
+	interruptionEvents, err := createSubscriptionClient(ctx, log, cfg.Project, cfg.PubSub.InstanceInterruptionSubscriptionName)
+	if err != nil {
+		log.Fatal("failed to init instance interruption subscription: %s", err.Error())
 	}
-	loggerConfig.EncoderConfig.TimeKey = "time"
-	logger, err := loggerConfig.Build()
+
+	creationEvents, err := createSubscriptionClient(ctx, log, cfg.Project, cfg.PubSub.InstanceCreationSubscriptionName)
 	if err != nil {
-		log.Fatalf("failed to initialize zap logger: %v", err)
+		log.Fatal("failed to init instance creation subscription: %s", err.Error())
 	}
-	sugar := logger.Sugar()
-	http.Handle(cfg.Prometheus.Path, promhttp.Handler())
-	go func() {
-		sugar.Fatal(http.ListenAndServe(fmt.Sprintf(":%s", cfg.Prometheus.Port), nil))
-	}()
 
-	n, err := createCSPNotifier(ctx, sugar, cfg)
+	computeClient, err := createComputeClient(ctx, log, cfg)
 	if err != nil {
-		sugar.Fatal("failed to init cloud-provider client: %s", err.Error())
+		log.Fatal("failed to init compute client")
 	}
 
-	e := make(chan events.InterruptionEvent)
-	go n.Receive(ctx, e)
-
-	c := cache.NewCacheWithTTL(time.Minute * 10)
-	for event := range e {
-		s := sugar.With("message_id", event.MessageID, "resource_id", event.ResourceID)
-		// this ensures we do not handle a duplicate message in the event pubsub sends it more than once
-		if exists := c.Exists(event.MessageID); exists {
-			s.Debug("handled duplicate message")
-			continue
-		}
-		c.Insert(event.MessageID)
-		interruptionEvents.WithLabelValues(cfg.ClusterName).Inc()
-		s.Info("interrupted")
+	initialInstances, err := computeClient.ListInstancesBelongingToKubernetesCluster(ctx)
+	if err != nil {
+		log.Fatal("failed to determine initial instances belonging to kubernetes clusters: %s", err.Error())
 	}
+
+	interruptions := make(chan *gcppubsub.Message, 30)
+	additions := make(chan *gcppubsub.Message, 30)
+	instanceToClusterMappings := cache.NewCacheWithTTLFrom(cache.NoExpiration, initialInstances)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
```

Comment:
What is this waitgroup doing? You're not passing it to anything, and not calling `wg.Done()` anywhere

Reply:
It is a hacky way to prevent the main goroutine from exiting. Behaviorally, there isn't much benefit in passing the waitgroup to each of the handlers because the only time they'll exit is when the channels which events are coming in from are closed, which would happen if there was a pubsub issue or the program is terminating. 

But for correctness yes, I'll pass the waitgroup to the two handlers